### PR TITLE
Add unit tests for DeleteCommand class #109

### LINKED ISSUE #109: Add unit tests for DeleteCommand class
as per title

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #109.

Ready for review.

### CODE REVIEW

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        DeleteCommand command = new DeleteCommand(targetVisibleIndex);
+        command.setData(addressBook, displayList);
+
+        return command;
+    }
+
+    /**
+     * Executes the command, and checks that the execution was what we had
+     * expected.
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Can be in one line? Our limit for line length is 110.

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        DeleteCommand command = createDeleteCommand(1, addressBook, emptyDisplayList);
+
+        assertCommandBehaviour(command, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX, addressBook,
+                addressBook);
+    }
+
+    @Test
+    public void execute_targetPersonNotInAddressBook_returnsPersonNotFoundMessage()
+            throws IllegalValueException {
+        Person notInAddressBookPerson = new Person(new Name("Alien"), new Phone("10000", false),
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Use realistic test data. e.g. "10000" is not a realistic phone number.

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    @Test
+    public void execute_indexMoreThanListSize_returnsInvalidIndexMessage() {
+        DeleteCommand command = createDeleteCommand(listWithSurnameDoe.size() + 1, addressBook,
+                listWithSurnameDoe);
+
+        assertCommandBehaviour(command, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX, addressBook,
+                addressBook);
+    }
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: See if you can put these three 'invalid index' cases in one method. All three has similar code, so possibly extract into a utility method?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    @Test
+    public void execute_validIndex_personIsDeleted() throws PersonNotFoundException {
+        final int targetPersonIndex = 0;
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Deleting the first person in the list is not very representative. Also try one index in the middle?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /**
+     * Executes the command, and checks that the execution was what we had
+     * expected.
+     * 
+     * @param deleteCommand to execute
+     * @param expectedMessage of the result of the execution
+     * @param expectedAddressBook of the result of the execution
+     * @param actualAddressBook that is modified by the command
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: These 4 lines don't seem to add much value. Delete?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        emptyDisplayList = new ArrayList<ReadOnlyPerson>();
+
+        listWithEveryone = new ArrayList<ReadOnlyPerson>();
+        listWithEveryone.add(johnDoe);
+        listWithEveryone.add(janeDoe);
+        listWithEveryone.add(davidGrant);
+
+        listWithSurnameDoe = new ArrayList<ReadOnlyPerson>();
+        listWithSurnameDoe.add(johnDoe);
+        listWithSurnameDoe.add(janeDoe);
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Consider creating a utility class e.g. `TestUtil` that provides methods to create test objects. e.g. `createAddressBook(Person... persons)`

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 7]
```diff
@@ -125,4 +125,11 @@ public void clear() {
         return internalList.iterator();
     }
 
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniquePersonList // instanceof handles nulls
+                && this.internalList.equals(
+                ((UniquePersonList) other).internalList));
```
Comment (damithc) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Probably can be put in the same line (if not longer than 110 chars).

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 8]
```diff
@@ -125,4 +125,11 @@ public void clear() {
         return internalList.iterator();
     }
 
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniquePersonList // instanceof handles nulls
+                && this.internalList.equals(
```
Comment (damithc) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: `||` and `&&` should not be aligned if they are not at the 'same level' in the condition tree. In this case, `||` is at a higher branch of the condition tree than `&&`.

**Comment by yamgent** [comment]
@damithc Thanks for your comments! I have made several changes wrt the review.

File: test/java/seedu/addressbook/util/TestUtil.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        
+        for (Person person : persons) {
+            list.add(person);
+        }
+        
+        return list;
+    }
+    
+    /**
+     * Creates an exact copy of the original address book.
```
Comment (damithc) [inline test/java/seedu/addressbook/util/TestUtil.java]: May want to mention how deep the copy is. e.g. is it going to contain same Person objects or copies of those Person objects?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        assertCommandBehaviour(command, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX, addressBook,
+                addressBook);
+    }
+    
+    /**
+     * Selects a person in the display list using the index, that does not exist in the address book, and 
+     * checks that the command reports that the person is not in the address book.
+     */
+    public void assertNoSuchPerson(int visibleIndex, AddressBook addressBook, 
+            List<ReadOnlyPerson> displayList) {
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: This indentation improves readability (2nd line starts after opening bracket of the first line and the closing bracket on the 2nd line is on the right of the end of first line)
```java
public void assertNoSuchPerson(int visibleIndex, AddressBook addressBook, 
                                                 List<ReadOnlyPerson> displayList) {
```

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        DeleteCommand command = createDeleteCommand(invalidVisibleIndex, addressBook, displayList);
+
+        assertCommandBehaviour(command, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX, addressBook,
+                addressBook);
+    }
+    
+    /**
+     * Selects a person in the display list using the index, that does not exist in the address book, and 
+     * checks that the command reports that the person is not in the address book.
+     */
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Method name is lacking some critical info. `assertDeletionFailsDueToNoSuchPerson()`?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    public void assertNoSuchPerson(int visibleIndex, AddressBook addressBook, 
+            List<ReadOnlyPerson> displayList) {
+
+        assert addressBook != null;
+        assert displayList != null;
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Perhaps no need for asserts as these are test methods. The test itself will fail if null is passed as parameters, right?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        
+        DeleteCommand command = createDeleteCommand(visibleIndex, addressBook, displayList);
+
+        assertCommandBehaviour(command, Messages.MESSAGE_PERSON_NOT_IN_ADDRESSBOOK, addressBook,
+                addressBook);
+    }
+    
+    /**
+     * Selects a person in the display list using the index, and tries to delete it. The expected outcome
+     * of the command should be successful.
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: `Asserts that the person at the specified index can be successfully deleted` ?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+     * 
+     * The addressBook passed in will not be modified (no side effects).
+     * 
+     * @throws PersonNotFoundException if the selected person is not in the address book
+     */
+    public void assertDeleteSuccessful(int targetVisibleIndex, AddressBook addressBook,
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: Does this need to be public?

File: test/java/seedu/addressbook/commands/DeleteCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        AddressBook expectedAddressBook = TestUtil.clone(addressBook);
+        expectedAddressBook.removePerson(targetPerson);
+
+        AddressBook actualAddressBook = TestUtil.clone(addressBook);
+        
+        DeleteCommand command = createDeleteCommand(targetVisibleIndex, actualAddressBook, displayList);
+
+        assertCommandBehaviour(command,
+                String.format(DeleteCommand.MESSAGE_DELETE_PERSON_SUCCESS, targetPerson), 
+                expectedAddressBook, actualAddressBook);
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/DeleteCommandTest.java]: This is neater
```java
String expectedMessage = String.format(DeleteCommand.MESSAGE_DELETE_PERSON_SUCCESS, targetPerson);

assertCommandBehaviour(command, expectedMessage, expectedAddressBook, actualAddressBook);
```

**Comment by yamgent** [comment]
@damithc I made further changes to the code as requested.

**Comment by damithc** [comment]
Good work @yamgent 
You may also create a PR to add yourself to the list of contributors.

**Comment by pyokagan** [comment]
@damithc Hmm, we may wish to set up gradle+checkstyle ASAP on this repo.

```
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:49: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:85: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:92: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:115: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:119: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:121: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:123: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:124: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:127: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:132: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:134: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:136: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:137: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:140: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:143: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:145: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:148: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:152: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:154: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:158: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/DeleteCommandTest.java:159: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:18: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:22: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:25: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:31: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:35: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/util/TestUtil.java:38: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
```

**Comment by damithc** [comment]
Indeed, it is good to have CI (including checkstyle) on level 3, 2, and even 1. Otherwise we are flying blind. The problem is there's no one to do it at the moment. @m133225 who was previously in charge of dev-ops stuff is not going to be very 'active' in this December.

**Comment by pyokagan** [comment]
@damithc I'm actually being quite overwhelmed right now myself since code reviews take a lot of time >< I'm hoping we can get more active developers on board :-D.

**Comment by damithc** [comment]
@pyokagan replied on slack

**Pull request merged**